build-system/tasks/build-bento.js

```diff
@@ -0,0 +1,271 @@
+ *
+ * Optionally copies the CSS at components/$name/$version/$name.css into
+ * a generated JS file that can be required from the components as
+ * `import {CSS} from '../../../build/$name-0.1.css';`
+ *
+ * @param {string} name Name of the extension. Must be the sub directory in
+ *     the components directory and the name of the JS and optional CSS file.
+ * @param {string} version Version of the extension. Must be identical to
+ *     the sub directory inside the extension directory
+ * @param {string} latestVersion Latest version of the extension.
+ * @param {boolean} hasCss Whether there is a CSS file for this extension.
+ * @param {?Object} options
+ * @param {!Array=} extraGlobs
+ * @return {!Promise<void>}
+ */
+async function buildComponent(
+  name,
+  version,
+  latestVersion,
+  hasCss,
+  options,
+  extraGlobs
+) {
+  options = options || {};
+  options.extraGlobs = extraGlobs;
+  if (options.compileOnlyCss && !hasCss) {
+    return;
+  }
+  const componentsDir = `src/bento/components/${name}/${version}`;
+  if (options.watch) {
+    await watchComponent(
+      componentsDir,
+      name,
+      version,
+      latestVersion,
+      hasCss,
+      options
+    );
+  }
+
+  if (hasCss) {
+    if (!existsSync('build')) {
+      mkdirSync('build');
+    }
+    if (!existsSync('build/css')) {
+      mkdirSync('build/css');
+    }
+    await buildExtensionCss(componentsDir, name, version, options);
+    if (options.compileOnlyCss) {
+      return;
+    }
+  }
+  if (options.npm) {
+    await buildNpmBinaries(componentsDir, name, options);
+    await buildNpmCss(componentsDir, options);
+  }
+  if (options.binaries) {
+    await buildBinaries(componentsDir, options.binaries, options);
+  }
+  if (options.isRebuild) {
+    return;
+  }
+
+  const bentoName = getBentoName(name);
+  await buildExtensionJs(componentsDir, bentoName, {
+    ...options,
+    wrapper: 'none',
+    filename: await getBentoBuildFilename(
+      componentsDir,
+      bentoName,
+      'standalone',
+      options
+    ),
+    // Include extension directory since our entrypoint may be elsewhere.
+    extraGlobs: [...(options.extraGlobs || []), `${componentsDir}/**/*.js`],
+  });
+}
+
+/**
+ * Builds a single component after extracting its settings.
+ * @param {!Object} components
+ * @param {string} component
+ * @param {!Object} options
+ * @return {!Promise<void>}
+ */
+async function doBuildComponent(components, component, options) {
+  const e = components[component];
+  const o = {...options, ...e};
+  await buildComponent(
+    e.name,
+    e.version,
+    e.latestVersion,
+    e.hasCss,
+    o,
+    e.extraGlobs
+  );
+}
+
+/**
+ * Build all the Bento components
+ *
+ * @param {!Object} options
+ * @return {!Promise<void>}
+ */
+async function buildBentoComponents(options) {
+  const startTime = Date.now();
+  maybeInitializeComponents();
+  const toBuild = getComponentsToBuild();
+  const results = [];
+  for (const component in components) {
```

Comment:
I think I may be missing something; where does `components` come from here?

Reply:
Disclaimer: This code is essentially a replica of that found in `build-system/tasks/extension-helpers.js` so while I can explain HOW it works, I really don't understand WHY it is written this way. It seems to be to prevent repeat calls to the method from duplicating work.

`components` is declared globally on line 24 and is then populated by the method `declareExctension` called by `maybeInitializeComponents`